Add mesh cap

DeniseWorthen opened the pull request:
# Pull Request Summary

Enables an alternate cap for use by UWM apps S2SW and HAFSW

## Description

The alternate source code utilizes the switch file ``switch_meshcap`` . A new switch option (``UWM``) is added to control features required by UWM.

The cap has a run-time config variable which determines whether the wave model is run in shel vs multi mode in UWM. Only a single grid has been verified when run in multi.

The new source files are:
- *wav_comp_nuopc*: This is the main cap code. It advertises and realizes connected fields on a provided ESMF mesh file. 
- *wav_import_export*: This fills import and export state fields.
- *wav_shel_inp*: Enables reading ``ww3_shel.inp`` file when run as shel.
- *wav_shr_mod*: A set of utility routines to reduce code duplication.
- *w3iogoncdmd.F90*: Writes gridded output fields as netcdf. Only a sub-set of all possible gridded output is currently available. When netcdf output is desired, an additional option needs to be added to the switch file (``UWMNCOUT``). This feature was originally coded to allow control via configuration variables but the code changes were backed out in order to resolve issues seen on the CESM side. A refactoring of the ``w3iogoncdmd.F90`` to allow all output variables is currently under development.
- *wav_kind_mod*

Code specific to CESM is added within ``ifdef CESMCOUPLED``. Code specific to UWM is added within ``ifdef W3_UWM``.

Please also include the following information: 
* Add any suggestions for a reviewer 
* Mention any labels that should be added:  _bug_, _documentation_, _enhancement_, _new feature_
* Are answer changes expected from this PR? Please describe the changes and the reason why in addition to which of the following labels would apply: _mod_def change_, _out_grd change_, _out_pnt change_, _restart file change_, _Regression test_ 
-->

### Issue(s) addressed
- fixes #636

### Commit Message

Adds source code to implement a mesh-based NUOPC cap in the dev/ufs-weather-model branch.

co-author: @mvertens

### Check list  

<!-- After creating the PR you can check each of the items below that have been completed -->

- [ ] Branch is up to date with the authoritative repository (NOAA-EMC) develop branch. 
- [ ] Checked the [checklist for a developer submitting to develop](https://github.com/NOAA-EMC/WW3/wiki/Code-Management#checklist-for-a-developer-submitting-to-develop). 
- [ ] If a version number update is required, checked the [updating version number](https://github.com/NOAA-EMC/WW3/wiki/Code-Management#checklist-for-updating-version-number) checklist. 
- [ ] If a new feature was added, a regression test for testing the new feature is added. 
 
### Testing

* How were these changes tested?
* Are the changes covered by regression tests? (If not, why? Do new tests need to be added?)
* Have the matrix regression tests been run (if yes, please note HPC and compiler)?
* Please provide the summary output of matrix.comp (_matrix.Diff.txt_, _matrixCompFull.txt_ and _matrixCompSummary.txt_):    
* Please indicate the expected changes in the regression test output ([Note the known list of non-identical tests](https://github.com/NOAA-EMC/WW3/wiki/How-to-use-matrix.comp-to-compare-regtests-with-master#4-look-at-results)).
* Please provide the summary output of matrix.comp (_matrix.Diff.txt_, _matrixCompFull.txt_ and _matrixCompSummary.txt_):

JessicaMeixner-NOAA committed:
more updating and clean up of make_makefile

JessicaMeixner-NOAA committed:
continue to make build script more modular

JessicaMeixner-NOAA committed:
fix typo

JessicaMeixner-NOAA committed:
Merge remote-tracking branch 'EMC/develop' into feature/expswitches

JessicaMeixner-NOAA committed:
update README in bin directory to be formatted

This displays on GitHub in a nice way

JessicaMeixner-NOAA committed:
Merge remote-tracking branch 'EMC/develop' into feature/cleanbuild

JessicaMeixner-NOAA committed:
add to bin/README

JessicaMeixner-NOAA committed:
update readme for formatting

JessicaMeixner-NOAA committed:
check in tools for ftn2src

JessicaMeixner-NOAA committed:
Merge remote-tracking branch 'EMC/develop' into feature/expswitches

 Conflicts:
	model/bin/make_makefile.sh
	model/ftn/w3gdatmd.ftn
	model/ftn/w3gridmd.ftn
	model/ftn/w3iogrmd.ftn
	model/ftn/w3srcemd.ftn
	model/ftn/w3wavemd.ftn

JessicaMeixner-NOAA committed:
Merge remote-tracking branch 'EMC/develop' into feature/expswitches

JessicaMeixner-NOAA committed:
Merge remote-tracking branch 'EMC/develop' into feature/expswitches

JessicaMeixner-NOAA committed:
Merge remote-tracking branch 'EMC/develop' into feature/expswitches

JessicaMeixner-NOAA committed:
remove variables and update mod_def date for xxx removal

JessicaMeixner-NOAA committed:
update w3gridmd for smc-- line should not be there

JessicaMeixner-NOAA committed:
remove two FLXX calls

JessicaMeixner-NOAA committed:
Merge remote-tracking branch 'EMC/develop' into feature/cleanbuild

JessicaMeixner-NOAA committed:
Merge remote-tracking branch 'origin/feature/expswitches' into feature/cleanbuild

JessicaMeixner-NOAA committed:
Merge remote-tracking branch 'EMC/develop' into feature/cleanbuild

JessicaMeixner-NOAA committed:
updates for build for recent commits

JessicaMeixner-NOAA committed:
Merge remote-tracking branch 'EMC/develop' into feature/cleanbuild

JessicaMeixner-NOAA committed:
update build with latest updates from develop which needed to be manually updated

JessicaMeixner-NOAA committed:
updating git ignore  list

JessicaMeixner-NOAA committed:
fix for build for NL5 and updating gitignore

JessicaMeixner-NOAA committed:
Updated version number to 7.13

JessicaMeixner-NOAA committed:
fix typo

JessicaMeixner-NOAA committed:
updates to sort_switch which allows w3_automake to work correctly

JessicaMeixner-NOAA committed:
updating reg and netcdf program list in w3_automake

JessicaMeixner-NOAA committed:
Merge remote-tracking branch 'origin/feature/cleanbuild' into version7.13

JessicaMeixner-NOAA committed:
committing testing build updates

JessicaMeixner-NOAA committed:
Merge remote-tracking branch 'EMC/develop' into feature/switch2cppWbuild_2

 Conflicts:
	manual/impl/compile.tex
	model/bin/make_makefile.sh
	model/bin/w3_make

JessicaMeixner-NOAA committed:
Merge remote-tracking branch 'origin/version7.13' into feature/switch2cppWbuild_2

JessicaMeixner-NOAA committed:
update make_makefile to correct version

JessicaMeixner-NOAA committed:
clean comments

JessicaMeixner-NOAA committed:
update readme based on feedback from Mickael

JessicaMeixner-NOAA committed:
Merge branch 'feature/cleanbuild' of github.com:JessicaMeixner-NOAA/WW3 into feature/cleanbuild

JessicaMeixner-NOAA committed:
Merge remote-tracking branch 'origin/feature/cleanbuild' into version7.13

JessicaMeixner-NOAA committed:
update about.tex to fill in previous v7 milestone descriptions

JessicaMeixner-NOAA committed:
remove temporary lines

JessicaMeixner-NOAA committed:
Merge remote-tracking branch 'origin/version7.13' into feature/switch2cppWbuild_2

JessicaMeixner-NOAA committed:
updates for using src instead of ftn

JessicaMeixner-NOAA committed:
update to use WWATCH_F90 in the cpp preprocessing

JessicaMeixner-NOAA committed:
removing un-needed setup files

JessicaMeixner-NOAA committed:
update ftn2src.sh

JessicaMeixner-NOAA committed:
update SCRIP makefile segments

JessicaMeixner-NOAA committed:
update script for conversion

JessicaMeixner-NOAA committed:
Merge remote-tracking branch 'EMC/develop' into feature/switch2cppWbuild_5

JessicaMeixner-NOAA committed:
fixing typos Chris found

JessicaMeixner-NOAA committed:
Merge remote-tracking branch 'EMC/develop' into feature/cleanbuild

JessicaMeixner-NOAA committed:
Merge remote-tracking branch 'EMC/develop' into feature/cleanbuild

JessicaMeixner-NOAA committed:
Merge remote-tracking branch 'origin/feature/cleanbuild' into version7.13

JessicaMeixner-NOAA committed:
Merge remote-tracking branch 'EMC/develop' into feature/cleanbuild

JessicaMeixner-NOAA committed:
Merge remote-tracking branch 'origin/feature/cleanbuild' into version7.13

JessicaMeixner-NOAA committed:
Merge remote-tracking branch 'EMC/develop' into version7.13

JessicaMeixner-NOAA committed:
merge commit

JessicaMeixner-NOAA committed:
Merge remote-tracking branch 'origin/version7.13' into feature/switch2cppWbuild_6

JessicaMeixner-NOAA committed:
updating source

JessicaMeixner-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/switch2cpp_build_nosrc

JessicaMeixner-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/switch2cpp_build_nosrc

JessicaMeixner-NOAA committed:
update ad3 to add a compiler specific statement

ukmo-ccbunney committed:
Provide CPP method of setting switch list in w3initmd

JessicaMeixner-NOAA committed:
Merge remote-tracking branch 'EMC/develop' into feature/switch2cpp_build_nosrc

JessicaMeixner-NOAA committed:
update src code with latest changes

JessicaMeixner-NOAA committed:
Merge branch 'feature/switch2cpp_0921code' into feature/switch2cpp_build_src

JessicaMeixner-NOAA committed:
cleaning up more unused files and updating .gitignore

JessicaMeixner-NOAA committed:
remove w3init switch part of ad3

JessicaMeixner-NOAA committed:
more ad3 clean-up

JessicaMeixner-NOAA committed:
clean up w3list is no longer built

JessicaMeixner-NOAA committed:
update comp.tmpl since we dont have w3list file anymore

JessicaMeixner-NOAA committed:
Merge branch 'feature/switch2cpp_build_nosrc' of github.com:JessicaMeixner-NOAA/WW3 into feature/switch2cpp_build_nosrc

JessicaMeixner-NOAA committed:
Revert "update comp.tmpl since we dont have w3list file anymore"

This reverts commit 176773d7d7a1b69627f8ff9f572ce5ee55393786.

JessicaMeixner-NOAA committed:
do not call w3list in comp.tmpl because we do not generate this exe anymore

JessicaMeixner-NOAA committed:
process src

JessicaMeixner-NOAA committed:
Merge branch 'update0922' into feature/switch2cpp_build_src

JessicaMeixner-NOAA committed:
first pass at ad3.tmpl file to get compiler specific files

JessicaMeixner-NOAA committed:
updates to make ad3.tmpl work

JessicaMeixner-NOAA committed:
missed section

JessicaMeixner-NOAA committed:
Merge remote-tracking branch 'EMC/develop' into feature/ad3template

JessicaMeixner-NOAA committed:
Merge remote-tracking branch 'EMC/develop' into feature/ad3template

JessicaMeixner-NOAA committed:
Merge remote-tracking branch 'EMC/develop' into feature/ad3template

JessicaMeixner-NOAA committed:
update ftn->src

JessicaMeixner-NOAA committed:
Merge remote-tracking branch 'origin/feature/ad3templatewithsrc' into feature/switch2cpp_build_src

JessicaMeixner-NOAA committed:
update for pgi

JessicaMeixner-NOAA committed:
add tools that convert ftn files with WW3 switch pre-processing
to src/*.F90 files with CPP switches
* This is the first step towards swtich->cpp conversion
* To use this script, from the tool directory run sh ftn2src.sh
* This processes the code 3 times so double switches are accounted for

JessicaMeixner-NOAA committed:
update what is going to ad3 for UKMO cray

JessicaMeixner-NOAA committed:
updates for UKMO for ad3

JessicaMeixner-NOAA committed:
add bin/bash and remove set-x

JessicaMeixner-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/tools_ftn2src

JessicaMeixner-NOAA committed:
add readme

JessicaMeixner-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/switch2cpp_build_nosrc

JessicaMeixner-NOAA committed:
Merge remote-tracking branch 'origin/feature/tools_ftn2src' into feature/switch2cpp_build_nosrc

Conflicts:
	model/tools/ftn2src.sh

JessicaMeixner-NOAA committed:
with processed source code

JessicaMeixner-NOAA committed:
Merge remote-tracking branch 'origin/temp/switch2cpp_build_nosrc' into feature/switch2cpp_build_src

Conflicts:
	model/src/gx_outp.F90
	model/src/w3gridmd.F90
	model/src/w3iogrmd.F90
	model/src/w3iopomd.F90
	model/src/w3odatmd.F90
	model/src/w3src4md.F90
	model/src/w3srcemd.F90
	model/src/w3wavemd.F90
	model/src/ww3_ounf.F90
	model/src/ww3_ounp.F90
	model/src/ww3_outp.F90

JessicaMeixner-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/switch2cpp_build_src

JessicaMeixner-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/switch2cpp_build_src

DeniseWorthen committed:
add ww3-cesm nuopc cap files

DeniseWorthen committed:
bring in Jessica's uncommitted build changes

DeniseWorthen committed:
add missing cpl_nuopc directory

DeniseWorthen committed:
make change in esmf/Makefile

DeniseWorthen committed:
tries to build shr_methods!

* copied cesm wav*F90 to src temporarily

DeniseWorthen committed:
building attempts

DeniseWorthen committed:
comment out memcheck in shr_methods

DeniseWorthen committed:
fix wave commit again

DeniseWorthen committed:
update bin/make_makefile.sh to not use cpl_nuopc subdir

DeniseWorthen committed:
update wav_import_export after fixing 'use, only'

DeniseWorthen committed:
remove extra files

DeniseWorthen committed:
working build

* use,only need to be added to wav_comp_nuopc

DeniseWorthen committed:
add use,only in wav_comp_nuopc

DeniseWorthen committed:
add optionally computed fields

Commit:
refactor of import/export routines to remove #ifdefs

DeniseWorthen committed:
cleanup

* add ST3 options in Calc routines
* remove extra CalcRoughl routine
* order Calc routines same as in wmesmfmod

DeniseWorthen committed:
add back state_fldchk routine

DeniseWorthen committed:
fix compile errors

Commit:
updates to wav_import_export to work with cesm

Commit:
new ifdefs for CESMCOUPLED

DeniseWorthen committed:
add cesmcoupled around stokes drift for cesm vs ufs

DeniseWorthen committed:
add wminitnml for non-cesm coupled

DeniseWorthen committed:
back-out addition of wminit attempt

Commit:
bug fixes and change of use of w3constants to constants

DeniseWorthen committed:
changes for ufs compile

* HML appears to be an added variable in cems' w3idatmd module
* current version of w3wave has non-optional input array odat(40)

DeniseWorthen committed:
remove some prints in shr methods

DeniseWorthen committed:
update w3wavemd w/ bug fix for deactivated sea points

DeniseWorthen committed:
updates for build/repro

* turn off wrst (debug flags still on) in switch file
* add cesm module to pass the runtype (need better fix)
* does not commit current wav_* versions; still need cleanup

DeniseWorthen committed:
initial commit for atm repro

DeniseWorthen committed:
tidy up

DeniseWorthen committed:
collect common fields for cesm and ufs

DeniseWorthen committed:
Merge branch 'dev/ufs-weather-model' into feature/w3capbs1

DeniseWorthen committed:
remove extra odat declaration

DeniseWorthen committed:
add stokes export

* global winds retrieved w/ correct name
* add stokes variables to exports

DeniseWorthen committed:
remove debugging line

DeniseWorthen committed:
change fill value for stokes

DeniseWorthen committed:
tidy up

DeniseWorthen committed:
initial repro w/ stokes export on

* z0 not repro but value exceeds max for atm use so
no impact

DeniseWorthen committed:
Merge remote-tracking branch 'origin/dev/ufs-weather-model' into feature/w3capbs1

DeniseWorthen committed:
restart repro with temp debugging code

* add reset of charn for activated/deactivated ice points
* ran both 6h and 12h restart cases, both passed
* use mod_def.mx100, created using glo_1deg grid.inp except
for turning off k-shifting (broke repro!)
* temp debugging code to be removed

DeniseWorthen committed:
remove debugging switches

DeniseWorthen committed:
remove un-used/needed items for ufs

* control restarts using odat ?? make here 3600s
* remove/comment out items that have no impact on ufs (why?)
* tidy up

DeniseWorthen committed:
reduce restart freq to 3h

DeniseWorthen committed:
fix inflags

* turning these off gives all zeros in exported fields

DeniseWorthen committed:
fix inflags, momentum

* set inflags(2:4) true, others false. Treat momentum (if used)
like wind in filling global data

DeniseWorthen committed:
bring in Matts debug changes 0096df6e

DeniseWorthen committed:
update cmplr.env

DeniseWorthen committed:
tidy up

DeniseWorthen committed:
fix errors shown by debug mode

DeniseWorthen committed:
Merge branch 'feature/w3capbs1_debug' into feature/w3capbs1

DeniseWorthen committed:
revert build files

DeniseWorthen committed:
add changes from w3capbs1 branch

DeniseWorthen committed:
Merge branch 'dev/ufs-weather-model' into feature/ww3capbuild

DeniseWorthen committed:
add back wcoss compile option and tidy up

DeniseWorthen committed:
replace wcoss compile options

DeniseWorthen committed:
another esmf/Makefile update

DeniseWorthen committed:
remove state_diagnose at realize, tidy up

* state_diagnose at realize failed in debug mode because of
sum() of uninitialized variables
* remove debug switches
* remove extra 'HK' print statements

DeniseWorthen committed:
more cleanup,alignment w/ mcapbuild,capbs1 branches

DeniseWorthen committed:
c96,c192,c384 tests all have b4b coupler restarts

DeniseWorthen committed:
tidy up

DeniseWorthen committed:
fix indent, remove netcdf q/dirty line

DeniseWorthen committed:
initial commit of wav_shel_inp

* model completes 6hrs in debug mode and writes out_pnt.points
and out_grd.mx100

DeniseWorthen committed:
prevent over-writing by local cesm values

* some variables should be used from the shel inp routine for
ufs but set locally for cesm
* restarts, grd and point output being written for ufs but
file name has no time string information

DeniseWorthen committed:
tidy up

DeniseWorthen committed:
sandbox testing for ww3 restart repro

*in debug mode, using dedicated ('second') output stream
produces non-identical ww3 restarts, but the model itself reproduces
this mode produces timestamped restart files

* in debug mode, using no data server mode, the ww3 restarts are
reproducible. this mode produces restarts restartXXX.ww3. base case
produces restarts at 09,012 (check in log.ww3), filenames restart001,
restart002. rest case produces only h=012, filename restart001.
cmp rest/restart001.ww3 base/restart002.ww3  shows the ww3 restarts
are b4b

DeniseWorthen committed:
fix time block for cesm

* move 'set time frame' block earlier to allow cesm
access to time and dtime_sync to set odat
* force odat(j=4) to be printed only once to log in
wav_shel_inp
* added "1" to end of point and grid out put spec in
ww3_shel.inp input file to force pre-pending by time

DeniseWorthen committed:
move ufs calls and make one cesmcoupled block

DeniseWorthen committed:
Merge branch 'feature/w3capbs3' into feature/ww3capbuild

DeniseWorthen committed:
remove scrip switches

DeniseWorthen committed:
fix two uninit faults using -init=snan,arrays

DeniseWorthen committed:
test branch for MV's latest changes

DeniseWorthen committed:
remove debugging code

DeniseWorthen committed:
remove more debugging code

DeniseWorthen committed:
update w/ hafs changes

DeniseWorthen committed:
update wav_comp_nuopc

DeniseWorthen committed:
clean up dbug statements and subnames

DeniseWorthen committed:
baselines pass with bs4 ww3 branch (c40a2762)

DeniseWorthen committed:
updates for merge_import

* remove flds_scalar_name from ToWave list, simplifies logic for
check_global SR
* set import_mask at start of import_fields using global reduction
and uwnd field from import state
* adds check_global routine for dumping out fields read from file,
the merged fields when merge_import is true and the import_mask

DeniseWorthen committed:
Merge branch 'feature/w3capbs4' of https://github.com/DeniseWorthen/WW3 into feature/w3capbs4

DeniseWorthen committed:
call set_importmask only for merge_import true

DeniseWorthen committed:
add comment, specify mdsf = ndsf(3)

* in ww3_shel.F90, ndsf(3) is assigned unit number 13 so use
that unit number in readfromfile

DeniseWorthen committed:
increase debug flag setting

* set dbug_flag to > 10 to write check_global

DeniseWorthen committed:
bring in MV's latest changes

DeniseWorthen committed:
compile fix

* remove import_mask from argument list

Commit:
added new cesmcoupled ifdefs to a subset of needed files

Commit:
updates for all CESMCOUPLED ifdefs

Commit:
fixed compile bug for ufs

Commit:
updates to get things running

DeniseWorthen committed:
updates to build and compile fix

* remove wav_wrapper_mod from build
* fix compile error

Commit:
updates to get cesm cases working out of the box

Commit:
updates to remove ww3.log file

Commit:
removed CESMCOUPLED if-defs that are no longer needed

Commit:
change needed for ufs build

Commit:
first set of changes to get debug mode working

DeniseWorthen committed:
add mesh mask check feature

* checks if mask in the provided mesh is all 1s (so
no valid land/sea mask) and replaces mesh mask with
internal land/sea mask

Commit:
fixed problems related to history writes

DeniseWorthen committed:
remove check for elementmask in mesh

mesh will contain an elementmask always, so don't need
to check if one exists

DeniseWorthen committed:
Merge branch 'dev/ufs-weather-model' into feature/ww3capbuild

DeniseWorthen committed:
Merge remote-tracking branch 'cesm/dev/unified' into feature/ww3capbuild

DeniseWorthen committed:
update bin and inp for ufs

DeniseWorthen committed:
Merge branch 'dev/ufs-weather-model' into feature/ww3capbuild

DeniseWorthen committed:
add SetVm and change log write from error to info

alperaltuntas committed:
add ww3a grid and IC files

alperaltuntas committed:
remove inadvertendly added __pycache__/ dir

alperaltuntas committed:
Merge pull request #15 from ESCOMP/add_ww3a

Add ww3a grid to dev/unified

DeniseWorthen committed:
modify setglobalinput

* do reduce across only sea points

DeniseWorthen committed:
modifications for nsea global reduce

* do reduce across only sea points
* add interface to do fill of global ww3 fields
* change variable name from data_global to global_data
* use ungridded dimension for wave elevation spectrum
* move util routines state_getfldptr and state_fldchk from wav_import_export to wav_shr mod

DeniseWorthen committed:
add interface to do fill of global ww3 fields

DeniseWorthen committed:
make merge_import work with nsea points

DeniseWorthen committed:
change variable name

rename data_global to global_data

DeniseWorthen committed:
clean up wav_coupling_to_cice arrays

* use ungridded dimension for wave_elevation_spectrum

DeniseWorthen committed:
compile fixes; move two SRS to wav_shr_mod

* move utility routines state_getldptr and state_fldchk from wav_import_export
to wav_shr_mod

DeniseWorthen committed:
cleanup wave_elevation_spectrum in wav_comp_nuopc

DeniseWorthen committed:
create multigrid branch of globalnsea

* add features from mcapstash

DeniseWorthen committed:
fix missing build changes; add comment

DeniseWorthen committed:
add missing multigrid conditional

DeniseWorthen committed:
move iounit setup for shel

* cleans up wav_comp_nuopc by creating new sr for opening
io units when not multigrid

DeniseWorthen committed:
add switch UWM

* add switch UWM for resetting charn in w3updtdmd for restart
repro

DeniseWorthen committed:
add switch and w3updtmd for uwm changes

DeniseWorthen committed:
add uwm netcdf gridded write changes

* optionally write netcdf grd output for uwm
* if history configuration is not set, will write native binary
ww3 gridded output

DeniseWorthen committed:
fix two compile errors

DeniseWorthen committed:
add missing time initialization for cesm

for multigrid, time variable cannot be set until after
initialization. Error is Attempt to use pointer TIME when it is not associated with a target
ufs sets time from time0 in read_shel_inp but cesm time is used to set odat so it must
be set prior to calling waveinit_cesm

Commit:
new #ifdefs introduced to not change the build for cesm

DeniseWorthen committed:
fix compile error

* all uwm baselines pass for both shel and multigrid

DeniseWorthen committed:
Merge remote-tracking branch 'origin/feature/globalnsea_multi' into feature/globalnsea_gridnc

DeniseWorthen committed:
revert uwm changes in w3wavemd

DeniseWorthen committed:
return uwm ifdef option to w3wavemd

DeniseWorthen committed:
remove cime_config and CESM CMakeLists.txt

DeniseWorthen committed:
Merge branch 'dev/ufs-weather-model' into feature/cmake_meshcapbuild

DeniseWorthen committed:
revert switch file and add swith_meshcap

kgerheiser committed:
Move switches to bin

kgerheiser committed:
Add option for multiple caps through the UFS_CAP option

Supported values are MULTI_ESMF and NUOPC_MESH

kgerheiser committed:
Test UFS_CAP options in workflows

DeniseWorthen committed:
Merge pull request #2 from kgerheiser/feature/cmake_meshcapbuild

Add new cap options to CMake

DeniseWorthen committed:
fix reversed logic for cap type

DeniseWorthen committed:
prepare merge w/ ufs cmake build for mesh cap

* remove cime_config and cesm CMakelist from branch committed
to dev/unified in order to merge cleanly with cmake mesh cap
changes on ufs side

DeniseWorthen committed:
Merge remote-tracking branch 'origin/feature/no_uwm_ncgrid' into feature/cmake_meshcapbuild

DeniseWorthen committed:
compile and grd.out fixes

* fix github action compile failure
* create new switch setting specifically to allow uwm netcdf
output, otherwise native binary output will be written. This is not
as flexible as initially implmented, which allowed netcdf gridded
output as a config flag. We reverted that ability in w3iogoncmd to
debug the perf issues on cesm side. UWM netcdf griddded output can
be turned on with compile-time switch UWMNCOUT
* all baselines pass run against shel_wavebl_20220214

alperaltuntas committed:
Merge pull request #18 from DeniseWorthen/feature/no_uwm_ncgrid

Implement use of NSEA points and time-axis in gridded output

JessicaMeixner-NOAA commented on model/src/wav_comp_nuopc.F90:
```diff
@@ -0,0 +1,1416 @@
+  !/
+  !/                  +-----------------------------------+
+  !/                  | WAVEWATCH III           NOAA/NCEP |
+  !/                  |           H. L. Tolman            |
```
Thanks for adding the WW3 header here.  The spot where you have "H.L. Tolman" should be the author(s) of this subroutine.  As I don't think Hendrik wrote this, so I think this needs to be updated.  In addition, I think the "last update:" date should be updated to today's or a recent date. 

A similar header should be included with every module/subroutine that is added.

DeniseWorthen replied:
It seems that the code documentation is being updated to utilize doxygen? If so, I think it makes more sense to move directly to inserting the required doxygen formatting.

JessicaMeixner-NOAA replied:
We are not removing the WW3 headers as part of the switch to doxygen.  Doxygen at this time is an addition.  Adding doxygen would be a welcomed addition and perhaps it could be a compromise for not adding the WW3 headers.  Perhaps we could still update this header to have proper authorship and dates if others are not added? @MatthewMasarik-NOAA can help if you have any questions about doxygen formatting.  We are not requiring doxygen headers since we haven't added doxygen to the rest of the code yet, but it would certainly be very welcomed to add now instead of coming back later to add it in.

DeniseWorthen replied:
I'm familiar w/ doxygen because I needed to add it for the grid generation code being added to ufs-utils. Does the documentation get built automatically for every push, as w/ ufs-utils?

MatthewMasarik-NOAA replied:
@DeniseWorthen We have not hooked in the build automation yet.  I'm actually working right now on a getting a current draft PR with the first batch of marked-up files ready for submission.  This will happen later today or tomorrow.  After that we'll be in a position to set up the auto build.

DeniseWorthen replied:
I've started adding the doxygen documentation to a separate branch which I will merge in to this PR branch when I've got the majority of it done. I can build it by hijacking the documentation build in my ufs-utils fork. 

To start, I used the ``ww3_doxy_tmpl.md`` file for the required features. What I found though is that if add both ``@param`` and then also ``!< `` to document the variables, it ends up "double documenting". For example, this code snip:

```
!> Determine if a field is in a state
!!
!! @param[in] State    an ESMF_State
!! @param[in] fldname  the name of the ESMF field
!!
!> @author mvertens@ucar.edu, Denise.Worthen@noaa.gov  @<01-05-2022>
  logical function state_fldchk(State, fldname)
    ! ----------------------------------------------
    ! Determine if field is in state
    ! ----------------------------------------------

    ! input/output variables
    type(ESMF_State) , intent(in)  :: State                  !< ESMF_State object
    character(len=*) , intent(in)  :: fldname                !< a field name
```

produces the following
<img width="780" alt="Screen Shot 2022-04-05 at 8 08 10 AM" src="https://user-images.githubusercontent.com/40498404/161751123-422bd94e-b4c5-49eb-b7d3-d29cc07dfe6d.png">

I wonder if it makes more sense to use @param only (or vice-versa) since documenting the arguments twice is redundant.

MatthewMasarik-NOAA replied:
Hi @DeniseWorthen , yes you're right, the two should not be used together. The `@param` should be used in headers for subroutines/functions, and the `!<` just for inline documenting of module variables.  All module variables are supposed to be documented.

MatthewMasarik-NOAA replied:
kudos for adding the doxygen markup!

DeniseWorthen replied:
@MatthewMasarik-NOAA Regarding module variables, do variables which are local only need to be docmented? That would mean things like ``integer :: i,j`` ?

MatthewMasarik-NOAA replied:
@DeniseWorthen , variables local to subroutines and functions don't need to be documented. Just variables that are listed between the 'module' and 'contains' keywords.  Does that answer it?

MatthewMasarik-NOAA replied:
When I said 'all' above, that's a little misleading ;)  only those non-locals.

kgerheiser committed:
Remove debug definition from UFS

A variable named DEBUG conflicts and is replaced by the value `1`.

DeniseWorthen committed:
Merge pull request #3 from kgerheiser/patch-1

Remove debug definition from UFS

Commit:
fixed dt settings so that they would not be overwritten by call to w3init

JessicaMeixner-NOAA commented on model/src/w3initmd.F90:
```diff
@@ -1236,6 +1246,9 @@ SUBROUTINE W3INIT ( IMOD, IsMulti, FEXT, MDS, MTRACE, ODAT      &
 !
       FLOUT(2) = NPT .GT. 0
 !
+#ifdef CESMCOUPLED
```
Delete commented out code?

JessicaMeixner-NOAA replied:
Not creating issue for this, assuming it will be taken care of in this PR.

JessicaMeixner-NOAA commented on model/src/w3updtmd.F90:
```diff
@@ -1803,6 +1807,9 @@ SUBROUTINE W3UICE ( VA )
 #endif
                 VA(:,JSEA) = 0.
+#if defined(W3_UWM) || defined(CESMCOUPLED)
+               charn(jsea) = aalpha
```
Curious why this and the below update was added?  I'm wondering if this should really be behind switches or just actual additions to the code?

DeniseWorthen replied:
I'm not clear on what you're referring to as a switch here? In my switch_meshcap file I have an item "UWM" to add this line only for UWM or CESM. How should I have added it?

The line was added as part of testing for restart repro.

JessicaMeixner-NOAA replied:
I'm wondering if instead of: 

```
#if defined(W3_UWM) || defined(CESMCOUPLED)
               charn(jsea) = aalpha
#endif      
```          

we should simply have: 
```
              charn(jsea) = aalpha
```

I'll run some tests to see the impact.

DeniseWorthen replied:
I was just trying to avoid modification of ww3 code for anything other than these two configurations.

JessicaMeixner-NOAA replied:
I'm testing this change: https://github.com/JessicaMeixner-NOAA/WW3/tree/bug/charnforrestart in the standalone WW3 regression tests as its my guess that this is a bug fix needed for all configurations.   I'll let you know how the tests go.

JessicaMeixner-NOAA replied:
I ran some tests and we do get differences, which is not unaccepted necessarily.  However, we it's taking me longer than expected to make sure there's are the diffs we want.  For now we can keep this behind the switches here, but I think we can make this more general in the very near future.

DeniseWorthen replied:
Thanks for testing; I'm behind on getting this and the UWM PR opened so I haven't addressed any comments yet. I needed a fix in the cap code and I've had trouble w/ the c192 test resources. I think I've resolved the issues and I'm testing hera and orion now.

JessicaMeixner-NOAA replied:
Is the 1/4 deg grid going to change at all?  If not, if it could be copied to hera (or another machine other than Cheyenne) I could in parallel start creating ICs.  Apologies if that's already been done and I've missed that.

DeniseWorthen replied:
There are two issues I think need to be addressed for the mod_def files. 

First, I am running w/o any obstr grid (they're all 0s).  At one point, Ali generated actual obstruction files to use when generating the mod_defs but I didn't incorporate them because I did not want to have to change the baseline I was testing against. 

Secondly, I created the grid.inp for the 1 deg tripole grid from the global 1-deg grid and for the 1/2 and 1/4 tripole grid from the 30m global grid. There are two things here which may need adjustment. First, the time-step settings and secondly the number of frequencies. I think for the actual prototypes, you may be using 36, but the RTs are using only 24 (?). 

The input-data on Hera is located at ``/scratch1/NCEPDEV/stmp4/Denise.Worthen/input-data-20220430/``. Within the ``WW3_input_data_20211113/createmoddefs`` are the 3 ww3_grid.inp.mx[100,050,025] that I have been using. 

Each of these lists a set of ascii files for input which I've generated from the fixed grid generation utility for the coupled model. These are all located at ``/scratch2/NCEPDEV/climate/Denise.Worthen/grids-esmf-20220116``

JessicaMeixner-NOAA replied:
The follow up for this issue is here: https://github.com/NOAA-EMC/WW3/issues/677

JessicaMeixner-NOAA reviewed (commented):
@DeniseWorthen thanks for making this PR! There's a lot of awesome updates here. I've made a few comments and am continuing to look at and review the code, but wanted to ask a few questions now. 
* I know this is a WIP, but I'm curious when we might be able to expect inputs or things available to run this + the ufs-weather-model on a platform other than Cheyenne, which will help in reviewing/testing the PR?  
* Should the CESMCOUPLED be W3_CESMCOUPLED? So it's just a "switch"? 

Again, thanks for all the awesome work and opening the draft PR.

DeniseWorthen commented:
> 

I've completed the update of my tests for waves+gocart now. I need to do a final round of updates for the actual tests. Once I do that, I can create a staged input data directory on hera. I know certain files will still need modifications---the mod_defs especially. It should probably be w/in the next day or so that I can have something workable on hera.

JessicaMeixner-NOAA commented on model/src/w3wavemd.F90:
```diff
@@ -1540,11 +1545,9 @@ SUBROUTINE W3WAVE ( IMOD, ODAT, TEND, STAMP, NO_OUT &
 ! 3.3.3 Update ice floe diameter
 !
-#ifdef W3_IS2
+#if defined(W3_IS2) || defined(CESMCOUPLED)
```
Why is just the IS2 switch not sufficient here?

mvertens replied:
@Jessica - CESM does not use W3_IS2 - and I did not write this code. I think this is a question for Cecelia Bitz or Lettie Roach who put this into the current ww3 code. @alperaltuntas - do you know the answer to this?

alperaltuntas replied:
I am not sure. I am mentioning @lettie-roach who may know the answer.

lettie-roach replied:
I don't think we need CESMCOUPLED here

JessicaMeixner-NOAA replied:
Assuming this will be updated in this PR and so no issues is made.

DeniseWorthen replied:
@JessicaMeixner-NOAA The combination of W3_IS2 and CESMCOUPLED has been removed.

alperaltuntas committed:
Merge pull request #20 from mvertens/feature/fix_dtsettings

fixed dt settings so that they would not be overwritten by call to w3…

DeniseWorthen committed:
use interface for state_getfldptr utility routine

DeniseWorthen committed:
remove use statements for ESMF and NUOPC

* use ESMF and use NUOPC mean the 'only' lines are redundant

DeniseWorthen committed:
update cesm timestep setting

* fix dt initialization for cesm

DeniseWorthen committed:
Merge remote-tracking branch 'remotes/ww3-cesm/dev/unified' into feature/cmake_meshcapbuild

DeniseWorthen committed:
remove cesm section

DeniseWorthen committed:
Merge remote-tracking branch 'origin/feature/fix2d' into feature/cmake_meshcapbuild

JessicaMeixner-NOAA commented on model/src/w3iogomd.F90:
```diff
@@ -1562,8 +1616,39 @@ SUBROUTINE W3OUTG ( A, FLPART, FLOUTG, FLOUTG2 )
             USSCO=FKD*SIG(IK)*WN(IK,ISEA)*COSH(2.*KD)
             BHD(JSEA) = BHD(JSEA) +                             &
                 GRAV*WN(IK,ISEA) * EBD(IK,JSEA) / (SINH(2.*KD))
+#ifdef CESMCOUPLED
```
All of these extra outputs would be nice to be available to all users instead of just behind the CESMCOUPLED ifdefs.  They would need to be updated in the manual with definitions and the template files for input files would need to be added along with a regression test testing the new output to ensure that they are tested and are not unintentionally broken by other users.  I understand that time will likely not allow for this to be done immediately, so perhaps we can make an issue for this to be addressed at a future time.

DeniseWorthen replied:
Agreed; this was meant to minimize any impact w/ the current code. Bringing in features like this for wider use would definitely be something we can work on.

DeniseWorthen replied:
The calculations require the surface layer depth HSL, set as 0.2* HML, where HML is the mixed layer depth imported from the Ocean. We don't currently export this field from the Ocean in UWM so this was kept behind the cesm ifdef.

JessicaMeixner-NOAA replied:
I'll have to look more but it seems like the "surface stokes drift" doesn't depend on HSL and we actually already have a surface stokes drift calculation (USSX,USSY) so understanding how those are different will be important at some point.  We might should add HML as an import field in the standalone mode to WW3 then?

mvertens replied:
@JessicaMeixner-NOAA - again I think this is a question for CC and Lettie. I'm not sure what their github ids are. @alperaltuntas - are you working on this particular part of the code base?

alperaltuntas replied:
These are the stokes drift calculations added by @qingli411, which are used in CESM's legacy OCN-WAV coupling that is based on passing the computed stokes drifts and langmuir enhancement factor to CVMix (through POP2 or MOM6). These changes were initially introduced to NCAR's old WW3 fork based on v.3.14.  @qingli411 can elaborate more.

qingli411 replied:
These Stokes drift calculations were added to enable the Langmuir turbulence parameterization in CESM through CVMix ([Li et al., 2016](https://doi.org/10.1016/j.ocemod.2015.07.020); [Li and Fox-Kemper 2007](https://doi.org/10.1175/JPO-D-17-0085.1)), which requires some scalar quantities such as the surface layer averaged Langmuir number and the Langmuir multiplier. These quantities can be diagnosed from the full profile of the Stokes drift. But they were added here in WW3 because (1) the full profile of Stokes drift is an array with depth rather than a scalar which is more expensive to pass from WW3 to the ocean model, and (2) the vertical resolution of the ocean model is too coarse to resolve the shape of the Stokes drift profile. 

@JessicaMeixner-NOAA The surface Stokes drift doesn't depend on the surface layer depth HSL (defined as 1/5 the mixed layer depth HML). It is calculated in newer versions of WW3 but was not there in the old version we had in CESM based on v3.14. Actually I think I ported this specific feature from v4.18. A difference in the calculation is that we added the contribution from a f^{-5} spectral tail, which seems to be commented out in newer versions of WW3. The surface layer averaged Stokes drift depends on HSL, but I'm not sure if it makes sense to add it as a standard WW3 output -- as you mentioned we will need to add HML as an import field for that but HML is not used in the calculation of any prognostic quantities in WW3. I think right now the use of these variables and the way they were implemented are really specific to CESM (or CVMix to be more accurate). Ideally, it would be nice to have a standard and easy way to pass the full Stokes drift profile to the ocean. Then these quantities can be diagnosed in the ocean model.

JessicaMeixner-NOAA replied:
@qingli411 if this does not make sense as an output of WW3 perhaps it should be moved to just a calculation of an import/export in the nuopc cap instead of being in the output computation routine?

qingli411 replied:
@JessicaMeixner-NOAA I think it does make sense to have these variables as output for CESM -- when running CESM coupled with WW3, it would be good to have some of them in the output. But I'm just not sure if it makes sense to add them to the standard output of WW3 if other users do not use them... So, perhaps it is OK to keep them behind the `CESMCOUPLED` ifdefs for now? In the future, I think it would be good to have a more general way to pass the full Stokes drift profile information to the ocean, e.g., something like what @breichl did? Then these quantities can be diagnosed from the Stokes drift profile in the ocean model.

JessicaMeixner-NOAA replied:
I think that's what's at issue here, we'd like for features that are not isolated to the cap to not be CESM only features and be available for non-CESM users.  If that requires adding an optional input as well, then we should add the optional input and make checks that say if you don't have x input then you can't have y output, as required.

JessicaMeixner-NOAA replied:
That being said, if this is temporary and we make an issue to remove this as you move to use the full Stokes profile, then perhaps just making an issue is fine.  However, then we have the issue that these are taking up places in the output array structures and so the next person who adds a variable is going to add something after that and then that gets messy to reconcile, so maybe this should use different indices so it'd be easier for others to add new variables in the meantime?

qingli411 replied:
@JessicaMeixner-NOAA OK I see. How about this? We can add a new standard output in WW3 for a depth averaged Stokes drift which requires a new import field of a depth. We don't need to specify the meaning of this depth, i.e., it can be 1/5 of the mixed layer depth as we've been using in CESM, but it can also be any depth of interest for other users. In the limit that this depth approaches zero, the depth averaged Stokes drift should be equivalent to the surface Stokes drift. In the limit that this depth approaches the water depth, the depth averaged Stokes drift should be the total Stokes transport divided by the water depth. I think both surface Stokes drift and total Stokes transport are standard output of WW3. This depth averaged Stokes drift can serve as another variable to describe the Stokes drift, which is potentially useful to other users. Then, I think the calculation of all other variables such as the Langmuir multiplier can be moved to the nuopc cap. What do you think?

There is another issue that whether we should include the contribution of a spectral tail to the Stokes drift. We are including the contribution from a f^{-5} spectral tail when using it in CESM. But I suspect that the Langmuir multiplier might not be very sensitive to this choice, so the effect of dropping this contribution is probably small...

JessicaMeixner-NOAA replied:
@qingli411 that sounds good to me.  

In terms of the tail to the Stokes drift, I'd recommend making an issue and then creating a flag to add it or not in a separate PR. My guess is others will be just as interested as you to have the option to add the tail or not.  But this way we only have one calculation of surface Stokes drift instead of multiple.

qingli411 replied:
@JessicaMeixner-NOAA Sounds good!

JessicaMeixner-NOAA replied:
The proposed subsequent updates are covered in these issues: 
* Add tail to USSX/USSY and do not add extra variable: https://github.com/NOAA-EMC/WW3/issues/669 
* Add "depth" input field. https://github.com/NOAA-EMC/WW3/issues/670 
* Move any relevant variable calculations to the cap code that do not need output: https://github.com/NOAA-EMC/WW3/issues/671
* Remove the ifdefs for any other remaining variables so that all users can access new output variables https://github.com/NOAA-EMC/WW3/issues/672

JessicaMeixner-NOAA commented on model/src/w3iogoncdmd.F90:
```diff
@@ -0,0 +1,759 @@
+#include "w3macros.h"
```
This is a really great feature.  I think this perhaps could inspire a larger conversation about how to optionally add in netcdf instead of binary output that also takes advantage of some of the recent work in ww3_ounf which allows for things like flexible user defined metadata and be usable by those who are not using CMEPS.

DeniseWorthen replied:
I currently have a feature branch which updates w3iogoncmd to write a netcdf file for any variable requested in the inp file. This current version was only set up to write a limited number of fields.

JessicaMeixner-NOAA replied:
An issue was created for this: https://github.com/NOAA-EMC/WW3/issues/673 where we can continue this discussion

JessicaMeixner-NOAA commented on model/src/w3iorsmd.F90:
```diff
@@ -430,18 +438,32 @@ SUBROUTINE W3IORS ( INXOUT, NDSR, DUMFPI, IMOD, FLRSTRT )
 !
 ! open file ---------------------------------------------------------- *
 !
+#ifdef CESMCOUPLED
```
Long term I don't think it's reasonable to support every systems custom naming conventions for restart files.  I hope that this can spur a conversation about how we could add some flexibility via namelist/input options that can add some customization meeting more users needs without system specific ifdefs.

JessicaMeixner-NOAA replied:
Would it be possible to find out the requirements for a resatrt file name for CESM so we can start a discussion about how to accommodate CESM in a more general way?

JessicaMeixner-NOAA replied:
Issue to address this is here: https://github.com/NOAA-EMC/WW3/issues/674

JessicaMeixner-NOAA commented on model/src/w3sic4md.F90:
```diff
@@ -388,8 +401,13 @@ SUBROUTINE W3SIC4 (A, DEPTH, CG, IX, IY, S, D)
+#ifdef CESMCOUPLED
+      ! It does not look like IC4PARS(1) is initialized anywhere - so
+      ! hard-wiring it here for now
+      IC4METHOD = 8
```
Based on the comment, perhaps this is a bug-fix that should be pulled out and tested without the #ifdef CESMCOUPLED unless IC4METHOD is a CESMFCOUPLED variable?

mvertens replied:
@lettie-roach - can you address this?

lettie-roach replied:
Hi @JessicaMeixner-NOAA -  I put this in because I couldn't figure out how to set this variable properly. I think it should be a namelist variable - it's not specific to CESMCOUPLED

JessicaMeixner-NOAA replied:
Have you tried setting in ww3_grid.inp: 
  &SIC4 IC4METHOD = 8 / 
?

lettie-roach replied:
Aha, no! We didn't have that file in the version we were using (https://github.com/lettie-roach/WW3-CESM/tree/lrmod, forked from Helen Kershaw). That is how it should be set

JessicaMeixner-NOAA replied:
An issue for this is created here: https://github.com/NOAA-EMC/WW3/issues/675

DeniseWorthen replied:
@JessicaMeixner-NOAA The line in question for hardcoding the ic4method has been removed.

JessicaMeixner-NOAA commented on model/src/w3sic4md.F90:
```diff
@@ -514,6 +532,50 @@ SUBROUTINE W3SIC4 (A, DEPTH, CG, IX, IY, S, D)
            WN_I= 0.5 * ALPHA
 
+#ifdef CESMCOUPLED
+        CASE (8) 
```
Is this something that could be added as an option that all can use?  This also seems like it could perhaps be a PR on it's own and is unrelated to the new cap although perhaps there is a dependency I do not fully understand here?

mvertens replied:
@JessicaMeixner-NOAA - I'm happy to try to set up a call with CC,Lettie and @alperaltuntas  if that would help.
Please let me know.

JessicaMeixner-NOAA replied:
@mvertens It would be great to get some answers to some of these questions.  If that would be easier in a call, then that would be great. Otherwise, I'm happy to continue via the GitHub interface.

mvertens replied:
@lettie-roach - can you answer the above question from Jessica?

lettie-roach replied:
We added case 8 as a new option for the attenuation scheme, so yes it's unrelated to the new cap. We have more changes to make to the attenuation scheme based on some recent results, so it would make sense to be a PR on its own

JessicaMeixner-NOAA replied:
An issue to remove ifdefs is here: https://github.com/NOAA-EMC/WW3/issues/676
I will leave the creation of a new issue about the additional changes to others.

JessicaMeixner-NOAA commented on model/src/w3timemd.F90:
```diff
@@ -77,6 +77,9 @@ MODULE W3TIMEMD
       LOGICAL, PRIVATE        :: FLPROF = .FALSE.
       CHARACTER, PUBLIC       :: CALTYPE*8
+#ifdef CESMCOUPLED
+      LOGICAL, PUBLIC         :: NOLEAP = .TRUE.
```
Not sure why the updates here are CESM specific other than not wanting to modify the non-cap related code?

lettie-roach replied:
I think Cecilia put in the leap-year related changes to fix a bug, so not CESM specific

JessicaMeixner-NOAA replied:
@cmbitz can you confirm if this should be pulled out a bug-fix and not something CESM specific?

DeniseWorthen replied:
I think this and the change at line 551 (below) are because the calendar type is initialized in the mod_def file instead of a namelist. I believe we can resolve this issue in the same way @mvertens resolved changing the timestep after the model initialization in the ``waveinit_cesm`` routine. But we will need to test and verify it works before removing this particular ifdef.

JessicaMeixner-NOAA replied:
I have not created an issue for this as I believe it will be taken care of in this PR.

DeniseWorthen replied:
@JessicaMeixner-NOAA The CESMCOUPLED ifdefs to hardcode noleap have been removed.

JessicaMeixner-NOAA commented on model/src/w3timemd.F90:
```diff
@@ -541,13 +548,17 @@ INTEGER FUNCTION MYMD21 ( NYMD )
       NM   = MOD(NYMD,10000) / 100
       ND   = MOD(NYMD,100)
+#ifdef CESMCOUPLED
+      LEAP=.false. ! seems to be not working when I made NOLEAP=.true., still getting leap day
```
Is there a bug that needs to be reported and addressed related to this comment?

DeniseWorthen replied:
I think the issue was that the CALTYPE is set during the creation of the mod_def. I'm not sure why it is there and not in an input namelist/inp file?

mvertens replied:
@JessicaMeixner-NOAA - I think we can get rid of this in the CESM initialization.

JessicaMeixner-NOAA commented on model/src/wav_shel_inp.F90:
```diff
@@ -0,0 +1,1028 @@
+
+  end subroutine set_shel_io
+
+  subroutine set_shel_inp(dtime_sync)
```
Based on the below settings, can we provide a ww3_shel.inp (or in the future a ww3_shel.nml) file so that we do not have this hard-coded input section?   This would allow flexibility to for example add an additional variable to the output without having to recompile the code and just modify an input file.

DeniseWorthen replied:
The set_shel_inp is only used by CESM (it is called as part of waveinit_cesm).

JessicaMeixner-NOAA replied:
Even if we are not using this, I still think it would be a nice feature for CESM to have that flexibility and would be cleaner code-wise.

JessicaMeixner-NOAA replied:
Follow up issue: https://github.com/NOAA-EMC/WW3/issues/678

JessicaMeixner-NOAA commented on model/src/wav_shel_inp.F90:
```diff
@@ -0,0 +1,1028 @@
+  end subroutine set_shel_inp
+
+  !===============================================================================
+  subroutine read_shel_inp(mpi_comm)
```
I don't think time will allow, but I'm thinking we should perhaps make this an independent subroutine that both ww3_shel.F90 would call as well the mesh based cap.  Additionally we can add in a future subroutine for ww3_shel.nml reading as well.  This way there would not be duplicative code.

DeniseWorthen replied:
I think breaking out the namelist/inp reading from the ww3_shel would make sense. Right now, what is in wav_shel_inp is probably too specific to the options used by UWM.

JessicaMeixner-NOAA replied:
Issue for this is here: https://github.com/NOAA-EMC/WW3/issues/679

JessicaMeixner-NOAA reviewed (commented):
@DeniseWorthen I know that you made the changes trying to minimize impacts to others which is appreciated.  However, I'm wondering if you have found a few bugs and added some great features that others would benefit from as well.  Due to time some of these items might have to become issues and addressed as a second step.

DeniseWorthen committed:
add doxygen formatting to mesh cap code

Squashed commit of the following:

commit 444212dd945f1d29135514723d5b62be38542863
Author: Denise Worthen <denise.worthen@noaa.gov>
Date:   Thu Apr 7 09:28:31 2022 -0400

    tidy up

commit 03435033b2aadc8b03791ef78bedbef2db5a5047
Author: Denise Worthen <denise.worthen@noaa.gov>
Date:   Thu Apr 7 09:05:42 2022 -0400

    more doxygen formatting added

commit 61c10acdf81b2718bd90717bb83560e236aa9a9c
Author: denise.worthen <Denise.Worthen@noaa.gov>
Date:   Wed Apr 6 22:17:19 2022 +0000

    updates more doxy

commit c9d10e4b851f1ffebece694db3de292ea9133e2d
Author: denise.worthen <Denise.Worthen@noaa.gov>
Date:   Wed Apr 6 22:02:51 2022 +0000

    fix doxy errors

    * still have warning that u_file_u is not documented ?

commit d19652f526f208ab23deee9266a037573cc75090
Author: Denise Worthen <denise.worthen@noaa.gov>
Date:   Wed Apr 6 17:15:34 2022 -0400

    more doxy

commit f1049d01d987fd325826c132ae2f37802dbd908e
Author: Denise Worthen <denise.worthen@noaa.gov>
Date:   Wed Apr 6 08:22:54 2022 -0400

    minor text placement changes

commit 24f99451b767c5fdc5a84fcb86610965963485ed
Author: Denise Worthen <denise.worthen@noaa.gov>
Date:   Tue Apr 5 15:51:47 2022 -0400

    update 2 more files

commit fe25b6321e3b2063cda4aa721f8eb6d5c31c08be
Author: Denise Worthen <denise.worthen@noaa.gov>
Date:   Tue Apr 5 12:58:03 2022 -0400

    initial commit of added doxygen

DeniseWorthen committed:
Merge branch 'dev/ufs-weather-model' into feature/cmake_meshcapbuild

JessicaMeixner-NOAA commented:
@DeniseWorthen in the wmesmf cap, the inputs are either from "cpl" or you can use the original method of inputs.  It looks like if we're using ww3_shel.inp that all inputs must come from the coupler?   Or is there another way that is determined?

DeniseWorthen commented:
@JessicaMeixner-NOAA I'm not clear on your question. Are you asking if the mesh cap can be run standalone or when the forcing fields come from files?

JessicaMeixner-NOAA commented:
> @JessicaMeixner-NOAA I'm not clear on your question. Are you asking if the mesh cap can be run standalone or when the forcing fields come from files?

So let's say we wanted to coupled atm-wav, so that winds are coming from the coupler, but we'd like to have an input field of ice that comes from a file and we are using ww3_shel.inp.  How would that get specified? Would that even be possible?

mvertens commented:
@JessicaMeixner-NOAA - with CDEPS inline functionality (so that a prognostic component can ingest forcing fields like you want and easily from its cap) - this could be easily done today. I'm happy to show how this can be implemented. I can go over this briefly at our upcoming call next week.

JessicaMeixner-NOAA commented:
> @JessicaMeixner-NOAA - with CDEPS inline functionality (so that a prognostic component can ingest forcing fields like you want and easily from its cap) - this could be easily done today. I'm happy to show how this can be implemented. I can go over this briefly at our upcoming call next week.

This is a capability we have with wmesmf when using ww3_multi.inp without using a separate data model.

mvertens commented:
@JessicaMeixner-NOAA - you are not using a separate data model - just share code functionality. I'm not suggesting this is the only way - but a powerful way to have different spatial and time interpolations (including 3d mapping) for forcing fields that it would be helpful for everyone to be aware of.

JessicaMeixner-NOAA commented on model/src/w3initmd.F90:
```diff
@@ -1402,6 +1415,16 @@ SUBROUTINE W3INIT ( IMOD, IsMulti, FEXT, MDS, MTRACE, ODAT      &
           CALL ALL_VA_INTEGRAL_PRINT(IMOD, "W3INIT, step 8.1")
 #endif
 #endif
+#ifdef CESMCOUPLED
```
Looking through this again, this seems like debug comments that are not needed long term.  Could this be cleaned out as well?

DeniseWorthen replied:
@JessicaMeixner-NOAA I have removed some sections where debugging print statements were added for CESMCOUPLED.

JessicaMeixner-NOAA reviewed (changes requested):
Issues to address review comments have been made and linked in the appropriate places.  In addition items that I believe will be taken care of in this PR were also noted.  If these items will not be addressed in this PR please let me know so an issue for that item can be created.

DeniseWorthen committed:
address pr review comments

* change cesmcoupled to w3_cesmcoupled
* removed commented out line in w3init
* mark sections which require fixes and testing
for cesm

DeniseWorthen committed:
address pr comments

* remove marked sections for pr review

DeniseWorthen committed:
remove debugging statements for CESMCOUPLED

JessicaMeixner-NOAA reviewed (commented):
All issues that were going to be taken care of before the PR was merged have been addressed.

DeniseWorthen committed:
Merge branch 'dev/ufs-weather-model' into feature/cmake_meshcapbuild

JessicaMeixner-NOAA reviewed (approved):
The regression tests are as expected: 

```
**********************************************************************
********************* non-identical cases ****************************
**********************************************************************
mww3_test_03/./work_PR2_UNO_MPI_d2                     (15 files differ)
mww3_test_03/./work_PR1_MPI_d2                     (16 files differ)
mww3_test_03/./work_PR3_UNO_MPI_d2_c                     (14 files differ)
mww3_test_03/./work_PR3_UQ_MPI_d2_c                     (15 files differ)
mww3_test_03/./work_PR3_UNO_MPI_d2                     (12 files differ)
mww3_test_03/./work_PR2_UQ_MPI_d2                     (15 files differ)
mww3_test_03/./work_PR3_UQ_MPI_d2                     (16 files differ)
ww3_ta1/./work_UPD0F_U                     (0 files differ)
ww3_tp2.10/./work_MPI_OMPH                     (7 files differ)
ww3_tp2.16/./work_MPI_OMPH                     (4 files differ)
ww3_tp2.17/./work_ma                     (1 files differ)
ww3_tp2.17/./work_a                     (1 files differ)
ww3_tp2.17/./work_mb                     (1 files differ)
ww3_tp2.17/./work_mc                     (1 files differ)
ww3_tp2.17/./work_c                     (1 files differ)
ww3_tp2.17/./work_b                     (1 files differ)
ww3_tp2.6/./work_ST0                     (1 files differ)
ww3_tp2.6/./work_ST4                     (1 files differ)
ww3_tp2.6/./work_pdlib                     (1 files differ)
ww3_ufs1.3/./work_a                     (1 files differ)
```
There are the known b4b cases + the unstructured mod_def files which have at times been not b4b lately.
